### PR TITLE
Add API to set thread affinity on Linux.

### PR DESCRIPTION
**Pull request by sharvil**
Issue: #2545

### CODE REVIEW

**Comment by sharvil** [comment]
Just a quick note on this PR – I changed the API a little bit so it's more flexible and matches `sched_setaffinity`.

**Comment by sharvil** [comment]
@martin-frbg, could you take a look at this PR? Thanks!

**Comment by martin-frbg** [comment]
LGTM, I'm just stuck on some of the other issues at the moment - will merge this later today. pthread_setaffinity_np seems to have come with glibc 2.3.4 in 2005 so we probably do not need a glibc version check around this for compatibilty. (Would be interested in your use case for this iff you can disclose it)

**Comment by sharvil** [comment]
Thanks a bunch, @martin-frbg!

In our case, we're building an ML inference library for a latency-sensitive model. That library can be linked in to various programs so using `taskset` on the entire process isn't great. We've also seen poor OS thread scheduling blowing away our latency budget. This change allows us to expose some tuning knobs that a) don't require OS-level changes and b) affect only our computations and not the rest of the process we're linked into.

File: driver/others/blas_server.c [hunk 1]
```diff
@@ -279,6 +280,23 @@ int get_node(void);
 
 static int increased_threads = 0;
 
+#ifdef OS_LINUX
```
Comment (h-vetinari) [inline driver/others/blas_server.c:283]: Shouldn't this use the same variable as the other if-guard? I.e. `OPENBLAS_OS_LINUX` instead of `OS_LINUX`?

Comment (martin-frbg) [reply driver/others/blas_server.c:283]: The API header for OpenBLAS has OPENBLAS_ prepended to the configuration variables,
internally i.e. for building OpenBLAS itself it is just OS_LINUX etc.

**Comment by mattip** [comment]
Could this get a mention in the wiki [extensions page](https://github.com/xianyi/OpenBLAS/wiki/OpenBLAS-Extensions)?

**Comment by martin-frbg** [comment]
Certainly, except that it is not yet available in a formal release.

**Comment by mattip** [comment]
Hmm. Is there a way to mark that as a post-release task?

File: driver/others/blas_server.c (hunk 1 above)
Comment (h-vetinari) [reply driver/others/blas_server.c:283]: thanks for the response, learned something new again.

**Comment by martin-frbg** [comment]
Hm, put a noose around my neck ? For the moment I hope to remember it when I create the changes document from the list of tagged tickets

File: driver/others/blas_server.c (hunk 1 above)
Comment (martin-frbg) [reply driver/others/blas_server.c:283]: come to think of it, this could probably be added to the extremely sketchy "developer manual" in the wiki

**Comment by mattip** [comment]
gentle ping about mentioning this in the wiki

**Comment by martin-frbg** [comment]
Hrrggh ? :flushed:
Added now, thanks for the reminder

**Pull request merged**